enhancements/network/ovn-kubernetes-ns-ipsec.md

```diff
@@ -0,0 +1,197 @@
+- Will not support policy to encrypt traffic between cluster nodes
+
+## Proposal
+
+Generally the propsal is to make libreswan available on the host and have the customer manage it's configuration via nmstate and k8s-nmstate. 
+For 4.14 TP, we'll simply use MCs with ipsec.config files.
+
+Becuase we can only have one pluto daemon on the node, E-W will have to move to use the host part too
+
+### 4.14
+
+this have the following parts:
+1. add libreswan and NM-libreswan to RHCOS as a new `ipsec` extension. We're adding both so that customer who need something that is not supported by the plugin can still achieve that by interacting directly with the lower libreswan layer.
+this alone is enough to enable n-s IPsec. customer can use MC to deploy certs and config files.
+
+2. But this is not enough. since there can be only one pluto daemon that listens on udp 500/4500, we also need to move E-W to the host. that is done by a series of changes to the ovn-ipsec daemonset:
+a. use host `/etc/ipsec.conf.d/openshift.conf` instead of the container `/etc/ipsec.conf`
+b. use hostPID so that pluto is actually "on the host". for that also mount relevant `/var/run` files.
+c. move the SIGTERM trap to a pod lifecycle preStop, because kubelet does not send SIGTERM when hostPID=true
+since the ipsec config file is now on the host, it doesnt really matter who start pluto as long as it's running. (ie - if customer configured NM-libreswan, and that kicks up way before ovn-ipsec container is started..)
+
+3. Upgrade. special care needed for upgrade, as CNO upgrade before the node, which means we wont have the os extension available. to avoid having to deal with it in 4.14 (see 4.15) - 
+a. we'll keep 2 sets of daemonsets `ipsec` and `ipsec-pre414` each daemonset containers (init, ipsec, and liveness) will check whether libreswan is installed on the host and act accordingly so that only one set is actually handling ipsec at a time
+b. this gives the added value of being able to mix nodes with n-s and without. still IMHO something undesireable in the future. we want to support a single implementation (ie - host based)
+
+### 4.15
+
+From 4.15, the os-extension and ipsec service will be configured through a new network.operator.openshift.io API that will allow to enable/disable ipsec and configure the e-w behaviour. the new API include the following options:
+Disabled = no ipsec. and is the default
+External = CNO will apply the os-extension to get libreswan on the nodes, enabling N-S ipsec
+Full = on top of external, CNO will also self-manage the E-W IPsec. E-W always imply also the possibility of user configurations for N-S. hence it is called `Full`.
+
+North-South configuration will be supported through nmstate (either via nmstate on the host or through k8s-nmstate operator).
+
+Upgrade will implement a state machine 
+
+For east-west only case, cno uses container based ipsec daemonset without deploying ipsec mc extension.
```

Comment:
this is no longer a valid statement for east-west only case. cno deploys ipsec mc extension and host based
ipsec daemonset.

Reply:
right!
it's left overs from 4.14!
but I believe we are still using the containers for HCP. right?

Reply:
need @pperiyasamy input here

Reply:
yes, it would be still container based ipsec daemonset for HCP due to unavailability of machine config support, for other cases it uses host based ipsec daemonset.